cellarium/ml/models/cellarium_gpt.py

```diff
@@ -0,0 +1,745 @@
+# Copyright Contributors to the Cellarium project.
+# SPDX-License-Identifier: BSD-3-Clause
+
+import math
+import random
+from collections.abc import Callable, Sequence
+from numbers import Number
+from typing import Any, Literal
+
+import lightning.pytorch as pl
+import numpy as np
+import torch
+import torch.distributions as dist
+from pykeops.torch import LazyTensor
+from torch import nn
+from torch.backends.cuda import sdp_kernel
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all
+
+from cellarium.ml.models.model import CellariumModel
+from cellarium.ml.utilities.testing import (
+    assert_arrays_equal,
+    assert_columns_and_array_lengths_equal,
+)
+
+
+class NegativeBinomial(dist.Distribution):
+    """Negative binomial distribution.
+
+    Args:
+        mu:
+            Mean of the distribution.
+        theta:
+            Inverse dispersion.
+    """
+
+    arg_constraints = {"mu": constraints.greater_than_eq(0), "theta": constraints.greater_than_eq(0)}
+    support = constraints.nonnegative_integer
+
+    def __init__(self, mu: torch.Tensor, theta: torch.Tensor, eps: float = 1e-8, validate_args: bool = False) -> None:
+        self.mu, self.theta = broadcast_all(mu, theta)
+        if isinstance(mu, Number) and isinstance(theta, Number):
+            batch_shape = torch.Size()
+        else:
+            batch_shape = self.mu.size()
+        self.eps = eps
+        super().__init__(batch_shape, validate_args=validate_args)
+
+    @property
+    def mean(self) -> torch.Tensor:
+        return self.mu
+
+    @property
+    def variance(self) -> torch.Tensor:
+        return self.mu + (self.mu**2) / self.theta
+
+    def log_prob(self, value: torch.Tensor) -> torch.Tensor:
+        if self._validate_args:
+            self._validate_sample(value)
+
+        # log_theta_mu_eps = torch.log(self.theta + self.mu + self.eps)
+        # return (
+        #     self.theta * (torch.log(self.theta + self.eps) - log_theta_mu_eps)
+        #     + value * (torch.log(self.mu + self.eps) - log_theta_mu_eps)
+        #     + torch.lgamma(value + self.theta)
+        #     - torch.lgamma(self.theta)
+        #     - torch.lgamma(value + 1)
+        # )
+        delta = torch.where(
```

Comment:
Add documentation:

"For large "theta", we can use Stirling's asymptotic approximation (see https://en.wikipedia.org/wiki/Gamma_function#Log-gamma_function), which is numerically more stable than PyTorch's implementation of lgamma."

Actually, the condition value / theta < 1e-2 may not be needed. Let's make an issue for me to investiagte.